packages/apollo-gateway/src/index.ts

```diff
@@ -343,40 +287,87 @@ export class ApolloGateway implements GraphQLService {
-          'Error checking for schema updates. Falling back to existing schema.',
-          e,
-        );
-        return;
-      }
+    if (options && options.engine) {
+      if (!options.engine.graphVariant)
+        console.warn('No graph variant provided. Defaulting to `current`.');
+      this.engineConfig = options.engine;
+    }
 
-      if (!result.isNewSchema) {
-        this.logger.debug('No changes to gateway config');
-        return;
-      }
+    let result: Await<ReturnType<Experimental_UpdateServiceDefinitions>>;
+    try {
+      this.logger.debug('Loading configuration for gateway');
+      result = await this.updateServiceDefinitions(this.config);
+    } catch (e) {
+      this.logger.warn(
+        'Error checking for schema updates. Falling back to existing schema.',
+        e,
+      );
+      return;
+    }
 
-      if (this.queryPlanStore) this.queryPlanStore.flush();
-      this.logger.debug('Gateway config has changed, updating schema');
+    if (!result.isNewSchema) {
+      this.logger.debug('No changes to gateway config');
+    }
 
-      this.schema = this.createSchema(result.serviceDefinitions);
-      try {
-        this.onSchemaChangeListeners.forEach(listener =>
-          listener(this.schema!),
-        );
-      } catch (e) {
-        this.logger.debug(
-          'Error notifying schema change listener of update to schema.',
-          e,
-        );
-      }
+    if (
+      JSON.stringify(this.serviceDefinitions) ===
+      JSON.stringify(result.serviceDefinitions)
+    ) {
+      this.logger.debug('No change in service definitions since last check');
+      return;
+    } else {
+      this.serviceDefinitions = result.serviceDefinitions;
+    }
+
+    this.compositionMetadata = result.compositionMetadata;
+    this.serviceDefinitions = result.serviceDefinitions;
+
+    if (this.queryPlanStore) this.queryPlanStore.flush();
+    this.logger.info('Gateway config has changed, updating schema');
+
+    this.schema = this.createSchema(result.serviceDefinitions);
+    try {
+      this.onSchemaChangeListeners.forEach(listener => listener(this.schema!));
+    } catch (e) {
+      this.logger.error(
+        'Error notifying schema change listener of update to schema.',
+        e,
+      );
+    }
+
+    if (this.experimental_didUpdateComposition) {
+      this.experimental_didUpdateComposition(
+        {
+          serviceDefinitions: result.serviceDefinitions,
+          schema: this.schema,
+          ...(this.compositionMetadata && {
+            compositionMetadata: this.compositionMetadata,
+          }),
+        },
+        previousServiceDefinitions &&
+          previousSchema && {
+            serviceDefinitions: previousServiceDefinitions,
+            schema: previousSchema,
+            ...(previousCompositionMetadata && {
+              compositionMetadata: previousCompositionMetadata,
+            }),
+          },
+      );
+    }
+  }
+
+  private startPollingServices() {
```

Comment:
could we similarly rename `stop`?

Reply:
`stop` is public - at best, we could deprecate and alias without making a breaking change for now.